### ISSUE TITLE
Async client method does not get executed until caller method returns

### ISSUE THREAD

**Issue by vuamitom** [facebook/fbthrift#307]
Hi, 

I'm putting together a small example of using FBThrift. There is one service with a single method that echo whatever number sent by client.

```
service ExampleService {
  i32 get_number(1:i32 number);  
}
```

Service implementation simply prints out and returns that number. 
```C++
class ExampleHandler: public ExampleServiceSvIf {
public:
	int32_t get_number(int32_t n) override {
		LOG(INFO) << "server: receive " << n;
		return n;
	}
};
```

I notice that the thrift client only sends request after the caller method have exited. For example, if call client method first and then put the thread to sleep for 10 sec, the server log shows that it receives request only after 10 sec duration ends. 

```C++
int main(int argc, char *argv[]) {
	LOG(INFO) << "Starting test ...";
	FLAGS_logtostderr = 1;
	folly::init(&argc, &argv);
        folly::EventBase eb;
	folly::SocketAddress addr("::1", thrift_port);
	auto client = newHeaderClient(&eb, addr);
        auto fut = client->future_get_number(16);
        std::move(fut).thenValue([](int32_t val) {
	       LOG(INFO) << "callback: " << val;
        });
        std::this_thread::sleep_for(std::chrono::milliseconds(10000));
        // server only prints out 16 after 10 sec
}
```

If I wrap the thrift client call in another method, the server receives request immediately without the 10sec wait.

```C++
void test() {
        folly::EventBase eb;
	folly::SocketAddress addr("::1", thrift_port);
	auto client = newHeaderClient(&eb, addr);
        auto fut = client->future_get_number(16);
        std::move(fut).thenValue([](int32_t val) {
	       LOG(INFO) << "callback: " << val;
        });
}

int main(int argc, char *argv[]) {
	LOG(INFO) << "Starting test ...";
	FLAGS_logtostderr = 1;
	folly::init(&argc, &argv);
        test();
        std::this_thread::sleep_for(std::chrono::milliseconds(10000));
        // server receives request immediately
}
```

Is the behavior expected? Shouldn't the client send request immediately when it's called instead of waiting until caller method returns? Or I am missing anything here?

**Comment by spalamarchuk** [comment]
@vuamitom, you're supposed to loop EventBase (via calling evb.loop() or evb.loopForever()) to make async models work. In your examples, you never do that, so it gets looped in its destructor just to drain existent events.

Note, that even in absence of OutputBufferingHandler, message cannot be guaranteed to be sent immediately because kernel buffers might be full or not big enough to accommodate the full message, in which case, several round-trips are required.

**Comment by vuamitom** [comment]
Yeah, thank you for pointing out. You're right. I had just looked through the generated code of sync call `client->sync_get_number` and realized that `loopForever` was expected after async call, and then replaced the call to `std::this_thread::sleep_for` above with `evb.loopForever` and then code worked. 

And thanks for mentioning the kernel buffers as well. I didn't think of it. But even in that case, it won't take as long as 10 sec for multiple round trips. In my case it's more like `std::this_thread::sleep_for` is blocking main thread and evenBase run loop can't start.
